packages/solidity-contracts/contracts/fuelchain/v2/FuelMessagePortalV2.sol

```diff
@@ -0,0 +1,145 @@
+    }
+
+    ////////////////////////
+    // Internal Functions //
+    ////////////////////////
+
+    /// @notice Performs all necessary logic to send a message to a target on Fuel
+    /// @param recipient The message receiver address or predicate root
+    /// @param data The message data to be sent to the receiver
+    function _sendOutgoingMessage(bytes32 recipient, bytes memory data) internal virtual override {
+        bytes32 sender = bytes32(uint256(uint160(msg.sender)));
+        unchecked {
+            //make sure data size is not too large
+            if (data.length >= MAX_MESSAGE_DATA_SIZE) revert MessageDataTooLarge();
+
+            // v2: increase deposited amount for sender
+            // Do not think this needs an overflow check, you cannot get that much ether
+            uint256 userDepositedAmount = depositedAmounts[msg.sender] + msg.value;
+            if(userDepositedAmount > depositLimitPerAccount) {
+                revert AccountDepositLimit();
+            }
+
+            depositedAmounts[msg.sender] = userDepositedAmount;
+
+            // v2: increase global deposited ether
+            uint256 globalDepositedAmount = totalDeposited += msg.value;
+            if(globalDepositedAmount > depositLimitGlobal) {
+                revert GlobalDepositLimit();
+            }
+            
+            //make sure amount fits into the Fuel base asset decimal level
+            uint256 amount = msg.value / PRECISION;
+            if (msg.value > 0) {
+                if (amount * PRECISION != msg.value) revert AmountPrecisionIncompatibility();
+                if (amount > type(uint64).max) revert AmountTooBig();
+            }
+
+
+            //emit message for Fuel clients to pickup (messageID calculated offchain)
+            uint nonce = _outgoingMessageNonce;
+            emit MessageSent(sender, recipient, nonce, uint64(amount), data);
+
+            // increment nonce for next message
+            _outgoingMessageNonce = nonce + 1;
+        }
+    }
+
+    /// @notice Executes a message in the given header
+    /// @param messageId The id of message to execute
+    /// @param message The message to execute
+    function _executeMessage(bytes32 messageId, Message calldata message) internal virtual override nonReentrant {
+        if (_incomingMessageSuccessful[messageId]) revert AlreadyRelayed();
+
+        //set message sender for receiving contract to reference
+        _incomingMessageSender = message.sender;
+
+        // v2: update accounting if the message carries an amount 
+        bool success;
+        bytes memory result;
+        if(message.amount > 0) {
+            uint256 withdrawnAmount = message.amount * PRECISION;
+            address recipient = address(uint160(uint256(message.recipient)));
+            uint256 depositedAmount = depositedAmounts[recipient];
+
+            if(depositedAmount < withdrawnAmount) {
+                depositedAmounts[recipient] = 0;
```

Comment:
Not sure I understand, are they trying to withdraw more than they have deposited? if that it true then this should error instead of letting them withdraw more

Reply:
You could be withdrawing to a different address, not the one you used originally to deposit.

I thought to myself that it might be worth it to completely remove this and not do accounting on withdrawal (save the global counter), just because of this issue, so that we save some gas. But I decided to do the naive implementation first

Reply:
Resolving

Reply:
hmm good point, maybe instead of a per account limit we should just cap the amount for each deposit to rate limit, and just stick to the global limit 🤔

Reply:
I am merging this one and opening another PR to showcase how the alternatives would look, accompanied with docs for the possible scenarios a user might face